### PR TITLE
wait all workers finish

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
ref https://github.com/pingcap/tikv/issues/1604
@BusyJay @siddontang @hhkbp2 PTAL

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -1902,12 +1903,24 @@ impl<T: Transport, C: PdClient> mio::Handler for Store<T, C> {
     // This method is invoked very frequently, should avoid time consuming operation.
     fn tick(&mut self, event_loop: &mut EventLoop<Self>) {
         if !event_loop.is_running() {
-            self.split_check_worker.stop();
-            self.region_worker.stop();
-            self.raftlog_gc_worker.stop();
-            self.compact_worker.stop();
-            self.pd_worker.stop();
-            self.consistency_check_worker.stop();
+            // Applying snapshot may take an unexpected long time.
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: add a start log here

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -1902,12 +1903,24 @@ impl<T: Transport, C: PdClient> mio::Handler for Store<T, C> {
...
+            handles.push(self.region_worker.stop());
+            handles.push(self.raftlog_gc_worker.stop());
+            handles.push(self.compact_worker.stop());
+            handles.push(self.pd_worker.stop());
+            handles.push(self.consistency_check_worker.stop());
+            for h in handles {
+                if let Some(h) = h {
+                    h.join().unwrap();
+                }
+            }
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: add a end log with total time here

**Comment by BusyJay** [comment]
LGTM

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -1902,12 +1903,24 @@ impl<T: Transport, C: PdClient> mio::Handler for Store<T, C> {
...
     fn tick(&mut self, event_loop: &mut EventLoop<Self>) {
         if !event_loop.is_running() {
-            self.split_check_worker.stop();
-            self.region_worker.stop();
-            self.raftlog_gc_worker.stop();
-            self.compact_worker.stop();
-            self.pd_worker.stop();
-            self.consistency_check_worker.stop();
+            // Applying snapshot may take an unexpected long time.
+            for peer in self.region_peers.values_mut() {
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: Extract to a stop method.

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -392,6 +393,31 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         Ok(())
     }
 
+    fn stop(&mut self) {
+        // Applying snapshot may take an unexpected long time.
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: add a start log here

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -392,6 +393,31 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        handles.push(self.compact_worker.stop());
+        handles.push(self.pd_worker.stop());
+        handles.push(self.consistency_check_worker.stop());
+        for h in handles {
+            if let Some(h) = h {
+                h.join().unwrap();
+            }
+        }
+
+        for peer in self.region_peers.values_mut() {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: add a end log with elapsed time here

**Comment by siddontang** [comment]
any update @zhangjinpeng1987 ?

**Comment by siddontang** [comment]
LGTM

**Pull request merged**